catroid/src/main/java/org/catrobat/catroid/web/ProjectDownloader.kt

```diff
@@ -0,0 +1,183 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.web
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import android.os.Bundle
+import android.os.Handler
+import android.os.ResultReceiver
+import android.support.annotation.VisibleForTesting
+import android.support.v7.app.AppCompatActivity
+import android.util.Log
+import org.catrobat.catroid.R.string.error_project_already_in_queue
+import org.catrobat.catroid.common.FlavoredConstants.DEFAULT_ROOT_DIRECTORY
+import org.catrobat.catroid.scratchconverter.Client.ProjectDownloadCallback
+import org.catrobat.catroid.transfers.project.ProjectDownloadService
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.ERROR_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.SUCCESSFUL_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.UPDATE_PROGRESS_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.UPDATE_PROGRESS_EXTRA
+import org.catrobat.catroid.ui.recyclerview.dialog.ReplaceExistingProjectDialogFragment
+import org.catrobat.catroid.utils.ToastUtil
+import org.catrobat.catroid.utils.notifications.StatusBarNotificationManager
+import java.io.File
+import java.io.Serializable
+import java.io.UnsupportedEncodingException
+import java.lang.ref.WeakReference
+import java.net.URLDecoder
+import java.util.Collections
+import java.util.HashSet
+import java.util.Locale
+
+class ProjectDownloader(
+    private val queue: ProjectDownloadQueue,
+    private val url: String,
+    callback: ProjectDownloadCallback?
+) : Serializable {
+    private val callbackWeakReference = WeakReference<ProjectDownloadCallback>(callback)
+
+    companion object {
+        private const val serialVersionUID = 42L
+        private const val FILENAME_TAG = "fname="
+        private val TAG = ProjectDownloader::class.java.simpleName
+
+        fun getProjectNameFromUrl(url: String): String? {
+            val projectNameIndex = url.lastIndexOf(FILENAME_TAG) + FILENAME_TAG.length
```

Comment:
please check out URLUtil from webkit package: guessFileName()

Reply:
Unfortunately this function doesn't work for our URL..